graph/src/data/graphql/validation_hash.rs

```diff
@@ -0,0 +1,415 @@
+use crate::prelude::q;
+use second_stack::buffer;
+use std::cmp::Ordering;
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+
+// The idea here is to avoid misses and increase the hit rate of the GraphQL validation's cache by normalizing GraphQL operations.
+// An exact same hash for two queries means that they are identical in the context of GraphQL Validation rules.
+//
+// Hashing includes:
+// - sorting of the fields, fragments, arguments, fragment definitions, variables etc.
+// - ignoring operation names
+// - transforming the selection sets like `query name { things }` into `{ things }`
+// - removing primitive values (like String, Int, Float except Boolean because it can change the body of the operation when used with `@include` or `@skip`)
+// - ignoring aliases
+pub fn validation_hash(query: &q::Document) -> u64 {
```

Comment:
This comment still needs to be addressed: [https://github.com/graphprotocol/graph-node/pull/3840#issuecomment-1217173718](https://github.com/graphprotocol/graph-node/pull/3840#issuecomment-1217173718) from [https://github.com/graphprotocol/graph-node/pull/3840#discussion_r947222988](https://github.com/graphprotocol/graph-node/pull/3840#discussion_r947222988).

The docs for `DefaultHasher` state:

> The internal algorithm is not specified, and so it and its hashes should not be relied upon over releases.

So, we don't know if it's at least a level 2 hash.

We can see from the source that `DefaultHasher::new()` hardcodes specific keys [https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#3155-3157](https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#3155-3157)
```rust
pub fn new() -> DefaultHasher {
    DefaultHasher(SipHasher13::new_with_keys(0, 0))
}
```

This means that an attacker can find collisions and use those collisions to harm an Indexer.

For example, an attacker could search for two queries with the same hash - one which is valid and the other invalid. They would serve the invalid query first, followed by the valid query. When the invalid query is served it would create an entry in `GRAPHQL_VALIDATION_CACHE` containing errors. The Indexer would correctly serve and attest to that result. Unfortunately, when the Indexer serves the second request it would incorrectly retrieve the validation errors because of the hash collision. It would then serve and attest to an error response for a query which should have validated. This can result in the indexer being slashed.

Using XXHash in the way suggested by the comment won't incur a significant performance penalty. Though, I should have pointed you to the streaming implementation here: [https://docs.rs/xxhash-rust/0.8.5/xxhash_rust/xxh3/struct.Xxh3Builder.html#method.with_secret](https://docs.rs/xxhash-rust/0.8.5/xxhash_rust/xxh3/struct.Xxh3Builder.html#method.with_secret)